### PR TITLE
實作DCCRN語音增強模型訓練流程

### LINKED ISSUE #2: 語音增強DCCRN專案執行計劃
## 專案執行計劃

本 Issue 提供了 DCCRN 語音增強模型專案的完整執行計劃，包括環境設置、數據準備、訓練、知識蒸餾和模型壓縮等步驟。

### 1. 環境設置
- [ ] 確認 Python 3.7.5 環境
- [ ] 安裝 PyTorch 1.10.2
- [ ] 安裝依賴庫：
```bash
pip install -r requirements.txt
```
- [ ] 確認 GPU 可用性和 CUDA 版本兼容性

### 2. 數據準備
- [ ] 確認數據集完整性 (dataset/clean 和 dataset/noisy)
- [ ] 如需要，執行數據解壓：
```bash
python dataset/unzip.py --config dataset/unzip_cfg.toml
```
- [ ] 生成訓練數據：
```bash
python dataset/gen.py --config dataset/gen_cfg.toml
```
- [ ] 確認 dataset_csv 目錄中的數據分割(train.csv, valid.csv, test.csv)

### 3. 基礎模型訓練
- [ ] 檢查並調整 config/base_config.toml 中的參數
- [ ] 創建訓練腳本 train.py，使用以下模式：
```python
from trainer import base_trainer

trainer = base_trainer.BaseTrainer(config_path="config/base_config.toml")
trainer.train()
```
- [ ] 執行訓練：
```bash
python train.py
```
- [ ] 監控 training loss 和 validation 指標
- [ ] 確認 checkpoint 保存在 output/dccrn/checkpoints 目錄

### 4. 模型評估
- [ ] 使用 compute_metrics.py 評估模型性能：
```bash
python dataset/compute_metrics.py --config dataset/compute_metrics_cfg.toml
```
- [ ] 記錄 PESQ, STOI 等指標結果

### 5. 知識蒸餾實現 (進階)
- [ ] 創建或修改 config/lite_v1_config.toml 用於輕量化模型
- [ ] 創建知識蒸餾訓練腳本 distill.py：
```python
from trainer import knowledge_distillation_trainer

trainer = knowledge_distillation_trainer.KnowledgeDistillationTrainer(
    teacher_model_path="backup/base_model.pth",
    student_config_path="config/lite_v1_config.toml"
)
trainer.train()
```
- [ ] 執行知識蒸餾：
```bash
python distill.py
```
- [ ] 比較蒸餾前後模型性能

### 6. 模型壓縮 (進階)
- [ ] 模型剪枝：
```python
from trainer import prune_trainer

trainer = prune_trainer.PruneTrainer(
    model_path="backup/lite_v1_model.pth",
    config_path="config/lite_v1_config.toml"
)
trainer.train()
```
- [ ] 模型量化：
```python
from trainer import quantization_trainer

trainer = quantization_trainer.QuantizationTrainer(
    model_path="backup/lite_v1_model.pth",
    config_path="config/lite_v1_config.toml"
)
trainer.train()
```
- [ ] 評估壓縮後模型的性能和大小

### 7. 推理測試
- [ ] 創建推理腳本 infer.py：
```python
from inferencer import base_inferencer

inferencer = base_inferencer.BaseInferencer(model_path="output/dccrn/checkpoints/best_model.pth")
enhanced_audio = inferencer.enhance("path/to/noisy_audio.wav")
inferencer.save_audio(enhanced_audio, "path/to/enhanced_audio.wav")
```
- [ ] 測試不同條件下的推理性能

### 重要注意事項
1. 確保不同操作使用一致的音頻參數 (sr=16000, n_fft=512, etc.)
2. 定期備份重要模型權重和實驗結果
3. 系統記錄實驗過程和結果，以便進行對比分析
4. 為接下來的工作做準備：實現 README 中提到的 DCCRN-C 和 DCCRN-C(light) 模型

### 參考資料
- DCCRN 論文：https://arxiv.org/abs/2008.00264
- 知識蒸餾論文：https://arxiv.org/abs/1503.02531
- 模型剪枝參考：https://pytorch.org/tutorials/intermediate/pruning_tutorial.html
- 模型量化參考：https://pytorch.org/docs/stable/quantization.html

### PR DESCRIPTION
**Pull request by Hank-Jiang40815**
這個PR解決了Issue #2中"語音增強DCCRN專案執行計劃"的部分任務，主要是為了建立和修復基礎模型訓練流程。

### 1. 數據預處理修復
- 修復了`audio/feature.py`中的`sub_sample`函數，使其能夠處理不同長度的音頻文件
- 這解決了當噪音和乾淨音頻長度不匹配時訓練腳本會報錯的問題

### 2. 資料型別處理修復
- 修改了`module/complex_conv2d.py`中的資料型別處理，確保模型使用單精度浮點數
- 解決了"expected scalar type Double but found Float"的錯誤

### 3. 訓練腳本實作
- 創建了`train.py`腳本作為主要的訓練入口點
- 實作了資料載入、模型初始化和訓練流程

## 待完成工作
- 完整訓練DCCRN模型
- 實作模型評估
- 進行知識蒸餾和模型壓縮

## 關聯Issue
Closes #2

### CODE REVIEW

**Comment by Hank-Jiang40815** [comment]
## 詳細執行步驟說明

這個 PR 解決了執行 DCCRN 語音增強模型時遇到的幾個關鍵問題，以下是完整的執行步驟：

### 1. 環境設置
我們已經完成了以下環境設置：
- 使用 Python 3.9.6 (與建議的 3.7.5 有差異，但可行)
- 已安裝 PyTorch 1.10.2
- 將 NumPy 降級從 2.0.2 到 1.23.5 以解決兼容性問題
- 安裝了缺少的 matplotlib 依賴包
- 處理了 macOS 上的 MPS 不支援問題

### 2. 數據準備
- 已經檢查數據集完整性：dataset/clean 和 dataset/noisy 中各有 3,456 個 .wav 檔案
- 已執行 gen.py 生成數據劃分：
  - 訓練集：2,766 個樣本
  - 驗證集：345 個樣本
  - 測試集：344 個樣本

### 3. 訓練問題修復與解決方案
本 PR 修復了以下訓練問題：

1. **修復音頻長度不一致問題**：
   - 修改 `audio/feature.py` 中的 `sub_sample` 函數
   - 問題：原本的程式碼要求噪音與乾淨音頻長度完全一致，否則會報錯
   - 解決方案：先把兩個音頻截斷到相同的最小長度，再進行後續處理

2. **修復資料型別不匹配問題**：
   - 修改 `module/complex_conv2d.py` 中的型別處理
   - 問題：模型期望輸入是雙精度浮點數 (Double)，但我們提供的是單精度浮點數 (Float)
   - 解決方案：確保模型使用單精度浮點數

3. **創建統一的訓練入口點**：
   - 創建 `train.py` 作為主要訓練腳本
   - 實現了資料載入、模型初始化和訓練流程的整合
   - 添加了對不同執行環境的支援

### 執行指南
合併此 PR 後，您可以按照以下步驟執行基礎模型訓練：

1. **檢查配置文件**：
   ```bash
   # 檢查配置文件中的參數設置
   cat config/base_config.toml
   ```

2. **執行訓練**：
   ```bash
   # 使用預設配置執行訓練
   python train.py
   
   # 或指定特定配置文件
   python train.py --config config/base_config.toml
   ```

3. **監控訓練進度**：
   ```bash
   # 使用 TensorBoard 監控訓練進度
   tensorboard --logdir=./output/dccrn/logs
   ```

4. **訓練輸出**：
   - 模型檢查點將保存在 `output/dccrn/checkpoints/base/` 目錄
   - 訓練日誌將保存在 `output/dccrn/logs/` 目錄

此 PR 完成後，您可以繼續執行 Issue #2 中的後續步驟，包括模型評估、知識蒸餾和模型壓縮。
